File: ios/SoboSociety/Models/SessionModels.swift
import Foundation

public struct ClassTypeDTO: Codable, Identifiable, Sendable {
    public let id: Int
    public let ad: String
    public let sure_dk: Int?
    public let renk: String?
    public let renk_kodu: String?

    public var displayColorHex: String {
        renk ?? renk_kodu ?? "#A2846F"
    }

    public init(id: Int, ad: String, sure_dk: Int? = 50, renk: String? = nil, renk_kodu: String? = nil) {
        self.id = id
        self.ad = ad
        self.sure_dk = sure_dk
        self.renk = renk
        self.renk_kodu = renk_kodu
    }
}

public struct InstructorDTO: Codable, Identifiable, Sendable {
    public let id: Int
    public let ad: String
    public let biyografi: String?
    public let foto_url: String?

    public init(id: Int, ad: String, biyografi: String? = nil, foto_url: String? = nil) {
        self.id = id
        self.ad = ad
        self.biyografi = biyografi
        self.foto_url = foto_url
    }
}

public struct ClassSessionDTO: Codable, Identifiable, Sendable {
    public let id: Int
    public let baslangic: String?
    public let baslangic_utc: String?
    public let bitis_utc: String?
    public let kontenjan: Int?
    public let toplam_kontenjan: Int?
    public let dolu_sayi: Int?
    public let doluluk: Int?
    public let durum: String?
    public let rezerve_edilebilir: Bool?
    public let uye_rezervasyonu_var: Bool?
    public let uye_bekleme_sirasinda: Bool?
    public let class_type: ClassTypeDTO?
    public let ders_tipi: ClassTypeDTO?
    public let instructor: InstructorDTO?
    public let egitmen: InstructorDTO?

    public var resolvedClassType: ClassTypeDTO {
        class_type ?? ders_tipi ?? ClassTypeDTO(id: 0, ad: "Reformer Pilates")
    }

    public var resolvedInstructor: InstructorDTO {
        instructor ?? egitmen ?? InstructorDTO(id: 0, ad: "Eğitmen")
    }

    public var resolvedKontenjan: Int {
        kontenjan ?? toplam_kontenjan ?? 10
    }

    public var resolvedDoluSayi: Int {
        dolu_sayi ?? doluluk ?? 0
    }

    public var kalanYer: Int {
        max(0, resolvedKontenjan - resolvedDoluSayi)
    }

    public var startFormatted: String {
        let isoStr = baslangic ?? baslangic_utc ?? ""
        if isoStr.contains("T") {
            let parts = isoStr.components(separatedBy: "T")
            if parts.count > 1 {
                let timeParts = parts[1].components(separatedBy: ":")
                if timeParts.count >= 2 {
                    return "\(timeParts[0]):\(timeParts[1])"
                }
            }
        }
        return isoStr
    }

    public init(
        id: Int,
        baslangic: String? = nil,
        baslangic_utc: String? = nil,
        bitis_utc: String? = nil,
        kontenjan: Int? = 10,
        toplam_kontenjan: Int? = 10,
        dolu_sayi: Int? = 0,
        doluluk: Int? = 0,
        durum: String? = "aktif",
        rezerve_edilebilir: Bool? = true,
        uye_rezervasyonu_var: Bool? = false,
        uye_bekleme_sirasinda: Bool? = false,
        class_type: ClassTypeDTO? = nil,
        ders_tipi: ClassTypeDTO? = nil,
        instructor: InstructorDTO? = nil,
        egitmen: InstructorDTO? = nil
    ) {
        self.id = id
        self.baslangic = baslangic
        self.baslangic_utc = baslangic_utc
        self.bitis_utc = bitis_utc
        self.kontenjan = kontenjan
        self.toplam_kontenjan = toplam_kontenjan
        self.dolu_sayi = dolu_sayi
        self.doluluk = doluluk
        self.durum = durum
        self.rezerve_edilebilir = rezerve_edilebilir
        self.uye_rezervasyonu_var = uye_rezervasyonu_var
        self.uye_bekleme_sirasinda = uye_bekleme_sirasinda
        self.class_type = class_type
        self.ders_tipi = ders_tipi
        self.instructor = instructor
        self.egitmen = egitmen
    }
}

File: android/lib/models/session_models.dart
class ClassTypeDTO {
  final int id;
  final String ad;
  final int kontenjan;
  final int sureDk;
  final String renk;

  ClassTypeDTO({
    required this.id,
    required this.ad,
    required this.kontenjan,
    required this.sureDk,
    required this.renk,
  });

  factory ClassTypeDTO.fromJson(Map<String, dynamic> json) {
    return ClassTypeDTO(
      id: json['id'] as int,
      ad: json['ad'] as String? ?? 'Ders',
      kontenjan: json['kontenjan'] as int? ?? 5,
      sureDk: json['sure_dk'] as int? ?? 50,
      renk: json['renk'] as String? ?? '#A2846F',
    );
  }
}

class InstructorDTO {
  final int id;
  final String ad;

  InstructorDTO({required this.id, required this.ad});

  factory InstructorDTO.fromJson(Map<String, dynamic> json) {
    return InstructorDTO(
      id: json['id'] as int,
      ad: json['ad'] as String? ?? 'Eğitmen',
    );
  }
}

class ClassSessionDTO {
  final int id;
  final String baslangic;
  final int kontenjan;
  final int doluSayi;
  final String durum;
  final double? fiyatTl;
  final ClassTypeDTO? classType;
  final InstructorDTO? instructor;

  ClassSessionDTO({
    required this.id,
    required this.baslangic,
    required this.kontenjan,
    required this.doluSayi,
    required this.durum,
    this.fiyatTl,
    this.classType,
    this.instructor,
  });

  factory ClassSessionDTO.fromJson(Map<String, dynamic> json) {
    return ClassSessionDTO(
      id: json['id'] as int,
      baslangic: json['baslangic'] as String,
      kontenjan: json['kontenjan'] as int? ?? 5,
      doluSayi: json['dolu_sayi'] as int? ?? 0,
      durum: json['durum'] as String? ?? 'active',
      fiyatTl: (json['fiyat_tl'] != null) ? (json['fiyat_tl'] as num).toDouble() : 900.0,
      classType: json['class_type'] != null ? ClassTypeDTO.fromJson(json['class_type']) : null,
      instructor: json['instructor'] != null ? InstructorDTO.fromJson(json['instructor']) : null,
    );
  }

  bool get isFull => doluSayi >= kontenjan;
  int get spotsLeft => (kontenjan - doluSayi).clamp(0, kontenjan);
}

class PackageDTO {
  final int id;
  final String ad;
  final int dersAdedi;
  final int gecerlilikGun;
  final double fiyatTl;
  final bool aktif;

  PackageDTO({
    required this.id,
    required this.ad,
    required this.dersAdedi,
    required this.gecerlilikGun,
    required this.fiyatTl,
    required this.aktif,
  });

  factory PackageDTO.fromJson(Map<String, dynamic> json) {
    return PackageDTO(
      id: json['id'] as int,
      ad: json['ad'] as String? ?? 'Ders Paketi',
      dersAdedi: json['ders_adedi'] as int? ?? 1,
      gecerlilikGun: json['gecerlilik_gun'] as int? ?? 30,
      fiyatTl: (json['fiyat_tl'] != null) ? (json['fiyat_tl'] as num).toDouble() : 0.0,
      aktif: json['aktif'] as bool? ?? true,
    );
  }
}
